### ISSUE TITLE
Consul ACL error with Vault 0.9.5

### ISSUE THREAD

**Issue by kronjob** [hashicorp/vault#4190]
Description :- Consul client throws ACL not found error with Vault 0.9.5. Same consul-access-token and same ACL policy works with Vault 0.7.

Reproduction steps :-
Running Vault 0.9.5 and Consul client in one kubernetes pod and consul server on another pod
storage "consul" {
address = "localhost:8500"
disable_registration = "true"
token = "$VAULT_CONSUL_KEY"
}
listener "tcp" {
address = "0.0.0.0:8200"
tls_disable = 1
}
listener "tcp" {
address = "0.0.0.0:9000"
tls_disable = 1
}
disable_mlock = true

Consul version :-
Client 1.0.6
Server 1.0.6

consul info for both Client and Server
Client:

agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease = 
	revision = 9a494b5f
	version = 1.0.6
consul:
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 1
	goroutines = 40
	max_procs = 1
	os = linux
	version = go1.9.3
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 25
	members = 5
	query_queue = 0
	query_time = 1
Server:

agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 0
build:
	prerelease = 
	revision = 9a494b5f
	version = 1.0.6
consul:
	bootstrap = false
	known_datacenters = 1
	leader = false
	leader_addr = 10.244.2.11:8300
	server = true
raft:
	applied_index = 742
	commit_index = 742
	fsm_pending = 0
	last_contact = 25.602µs
	last_log_index = 742
	last_log_term = 2
	last_snapshot_index = 0
	last_snapshot_term = 0
	latest_configuration = [{Suffrage:Voter ID:bddb8ded-00bc-a865-b090-3f1c8304bbf1 Address:10.244.2.11:8300} {Suffrage:Voter ID:f1316e8a-0ba4-02ec-6c17-d09ae5ed8274 Address:10.244.1.25:8300} {Suffrage:Voter ID:16261898-9135-803f-c989-26a34eae74c8 Address:10.244.0.62:8300}]
	latest_configuration_index = 1
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 2
runtime:
	arch = amd64
	cpu_count = 1
	goroutines = 77
	max_procs = 1
	os = linux
	version = go1.9.3
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 2
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 25
	members = 5
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 5
	members = 3
	query_queue = 0
	query_time = 1

**Comment by jefferai** [comment]
Did the Consul version change as well? Consul changed their ACL system quite a lot a bit ago.

**Comment by kronjob** [comment]
No, I kept both Consul client and server at 1.0.6 even Consul ACL policy stays same. Only thing I changed was Vault from 0.7.0 to 0,9.5

**Comment by jefferai** [comment]
I asked the Consul team for ideas, here's what they said:
> I would ask them to verify that the token its complaining about exists. If they have a management token to consul they can use https://www.consul.io/api/acl.html#list-acls to see if its there. they are running everything in k8s from what I see on the ticket so its hard to know if the data dir changed out from underneath consul when they upgraded.

**Comment by kronjob** [comment]
Thanks jefferai for your help. I will check token existence in ACL list and get back on this with more information. I am new to this Vault+Consul thing and exploring this integration for our use-case. 
Between the two scenarios(Vault 0.7 and Vault0.9), I was bringing up a new kubernetes cluster each time so data dir for consul is newly created. Additionally, Vault0.9 with Consul1.0.6 works without ACL for me. I will share more details after some time.

**Comment by kronjob** [comment]
Hi Jefferai,

Sorry for delay. Got stuck with some other job.
I did a run again with Consul(1.0.6) and Vault(0.9.5)+Consul-agent-client(1.0.6).

The token used for Vault's access can be seen in Consul UI.


![image](https://user-images.githubusercontent.com/37552205/38175356-f4d7032c-35f8-11e8-9d76-4c6c061a5abe.png)

**Comment by kronjob** [comment]
Below are some logs from **consul-agent-client** at the time of Vault's query.Please see the bold text.

2018/03/31 17:20:55 [ERR] consul: "**KVS.Get" RPC failed to server 10.244.0.23:8300: rpc error making call: rpc error making call: ACL not found**
    2018/03/31 17:20:55 [DEBUG] manager: cycled away from server "consul-1"
    2018/03/31 17:20:55 **[ERR] http: Request GET /v1/kv/vault/core/seal-config, error: rpc error making call: rpc error making call: ACL not found from=127.0.0.1:37190**
    2018/03/31 17:20:55 [DEBUG] http: Request GET /v1/kv/vault/core/seal-config (1.064001ms) from=127.0.0.1:37190
    2018/03/31 17:21:00 [DEBUG] Skipping remote check "serfHealth" since it is managed automatically
    2018/03/31 17:21:00 [DEBUG] agent: Node info in sync
    2018/03/31 17:21:04 [DEBUG] memberlist: Stream connection from=10.244.0.22:41692
    2018/03/31 17:21:14 [DEBUG] memberlist: Stream connection from=10.244.0.24:48592
    2018/03/31 17:21:24 [DEBUG] memberlist: Initiating push/pull sync with: 10.244.0.22:8301
    2018/03/31 17:21:44 [DEBUG] memberlist: Stream connection from=10.244.0.24:48690
    2018/03/31 17:21:54 [DEBUG] memberlist: Initiating push/pull sync with: 10.244.1.9:8301
    2018/03/31 17:21:54 [DEBUG] memberlist: Stream connection from=10.244.1.9:41292
    2018/03/31 17:22:14 [DEBUG] memberlist: Stream connection from=10.244.0.24:48778

**Comment by kronjob** [comment]
Next I tried directly (using both master token and Vault token) with curl from vault-agent-client(container) 
I get no response as the key-value pair doesn't exist in Consul yet.

Vault client token -->
/ # curl --request GET --header "X-Consul-Token: e8f72cdc-3506-11e8-b85d-13ed0be0e243" http://10.244.0.23:8500/v1/kv/vault/core/seal-config

Master token -->
/ # curl --request GET --header "X-Consul-Token: 74e1166e-3506-11e8-9431-27e752a263da" http://10.244.0.23:8500/v1/kv/vault/core/seal-config

Next I inserted a dummy key-value pair at vault/core/ using consul CLI : -

$consul kv put vault/core/seal-con 101
Success! Data written to: vault/core/seal-con

$consul kv get -detailed vault/core/seal-con
CreateIndex      183
Flags            0
Key              vault/core/seal-con
LockIndex        0
ModifyIndex      183
Session          -
Value            101

Then I queried this key using above tokens from within Vault-agent-client container : -

Master token -->
/ # curl --request GET --header "X-Consul-Token: 74e1166e-3506-11e8-9431-27e752a263da" http://127.0.0.1:8500/v1/kv/vault/core/seal-con
[{"LockIndex":0,"Key":"vault/core/seal-con","Flags":0,"Value":"MTAx","CreateIndex":183,"ModifyIndex":183}]/ #

Vault token -->
/ # curl --request GET --header "X-Consul-Token: e8f72cdc-3506-11e8-b85d-13ed0be0e243" http://10.244.0.23:8500/v1/kv/vault/core/seal-con
[{"LockIndex":0,"Key":"vault/core/seal-con","Flags":0,"Value":"MTAx","CreateIndex":183,"ModifyIndex":183}]/ #

**Comment by kronjob** [comment]
/etc/consul/server.json : - 
I kept default policy to "allow" for debugging (temporarily)
{
  "ca_file": "/etc/tls/ca.pem",
  "cert_file": "/etc/tls/consul.pem",
  "key_file": "/etc/tls/consul-key.pem",
  "verify_incoming": true,
  "verify_outgoing": true,
  "verify_server_hostname": true,
  "acl_datacenter": "dc1",
  "acl_default_policy": "allow",
  "acl_master_token": "74e1166e-3506-11e8-9431-27e752a263da",
  "acl_agent_master_token": "74e15c1e-3506-11e8-a577-5ffcf7810efe",
  "acl_agent_token": "74e19d0a-3506-11e8-8bca-efb980c1ae6f",
  "ports": {
    "https": 8443
  }
}

**Comment by kronjob** [comment]
I am unsure whether this is a Vault issue or Consul client. It seems to be Consul client issue to me or maybe something wrong with Vault's request to Consul client(maybe absent ACL token). As I mentioned earlier, same setup works for Vault 0.7(with ACL) or Vault 0.9.5(without ACL). I wrote so many comments to segregate logs logically for better readability.  Please let me know if any other information can help in debugging this issue.

**Comment by jefferai** [comment]
This seems like the key to me:

2018/03/31 17:20:55 [ERR] consul: "KVS.Get" RPC failed to server 10.244.0.23:8300: rpc error making call: rpc error making call: ACL not found

It's telling you the Consul token wasn't found...my guess is that you're setting it one place doing the query but not another? 

I don't really follow what the various tokens and clients are in the output above, especially since from the output in https://github.com/hashicorp/vault/issues/4190#comment-377799884 it seems like things are working?, but that error message is quite clear.

**Comment by kronjob** [comment]
"It's telling you the Consul token wasn't found...my guess is that you're setting it one place doing the query but not another?"

Yeah..that's the mystery for now. The same token which my Vault+ConsulClient is supposed to use is visible in Consul UI(as Vault-token) and also working using curl client.
Why I posted in this forum is everything works for me with Vault0.7 but not in 0.9.5.
I have also posted in consul forum but no response. 
https://github.com/hashicorp/consul/issues/3972
